src/librustc_trait_selection/traits/project.rs

```diff
@@ -1258,6 +1265,42 @@ fn confirm_generator_candidate<'cx, 'tcx>(
         .with_addl_obligations(obligations)
 }
 
+fn confirm_discriminant_kind_candidate<'cx, 'tcx>(
+    selcx: &mut SelectionContext<'cx, 'tcx>,
+    obligation: &ProjectionTyObligation<'tcx>,
+    discriminant_kind_vtable: VtableDiscriminantKindData<'tcx, PredicateObligation<'tcx>>,
+) -> Progress<'tcx> {
+    let tcx = selcx.tcx();
+
+    let enum_ty = selcx.infcx().shallow_resolve(discriminant_kind_vtable.enum_ty);
+    let substs = tcx.mk_substs([enum_ty.into()].iter());
+
+    let assoc_items = tcx.associated_items(tcx.lang_items().discriminant_kind_trait().unwrap());
+    // FIXME: emit an error if the trait definition is wrong
+    let discriminant_def_id = assoc_items.in_definition_order().next().unwrap().def_id;
+
+    let discriminant_ty = match enum_ty.kind {
+        // Use the discriminant type for enums
+        ty::Adt(adt, _) if adt.is_enum() => adt.repr.discr_type().to_ty(tcx),
+        // Default to `i32` for generators.
+        //
+        // FIXME: We may want to use the smallest
+        // fitting integer type instead.
+        ty::Generator(..) => tcx.types.i32,
+        // Other types do not have actually need a discriminant, so
+        // we can use a ZST to save space.
+        _ => tcx.types.unit,
```

Comment:
We should make sure this is compatible with the MIR `Rvalue::Discriminant`'s type. That one might be using `u8` as a fallback, in which case you should too, here.

Reply:
Actually, `u8` might remove some special-cases, since it's an integer and all non-`enum`/generator types always have discriminant `0` already.

Reply:
:thinking: We are currently incompatible by always using `u64`.

I would rather like to change `Rvalue::Discriminant` to a zst instead (which shouldn't be done in this PR). Using a `ZST` for `discriminant_value` should greatly improve codegen for single variant enums: https://rust.godbolt.org/z/M7iyak